### DISCUSSION TITLE
Loading models twice

### DISCUSSION THREAD

**Discussion by talelmishali** [spatie/laravel-permission#1839] [Q&A]
Hello,
I am passing an array with the user ```roles``` and ```permissions```  back with the request to the client for authorization on the front end. 
```php
    public function share(Request $request)
    {
        return array_merge(parent::share($request), [
            'authorization' => [
                'roles' => request()->user()?->roles->pluck('name'),
                'permissions' => request()->user()?->getAllPermissions()->pluck('name'),
            ],
        ]);
    }
```

And using ```roles``` and ```permissions``` on the ```Controller```policies as well, the end result is that I load the ```Spatie\Permission\Models\Permission``` & ```Spatie\Permission\Models\Role``` twice. 

Any good idea how to avoid loading it twice and save memory?

Thank you,
Tal

**Comment by erikn69** [comment]
I did not understand your question, if you mean running two sqls, one for roles and one for permissions, that's the way laravel handles relationships and that does not belong in this package, use `request()->user()?->permissions->pluck('name')` instead

**Reply by erikn69** [reply]
@talelmishali maybe i did understand now,  you mean, there are running four sqls, two for roles and two for permissions
Try `auth()->user()?->permissions->pluck('name');` and `auth()->user()?->roles->pluck('name');`

**Reply by talelmishali** [reply]
I am using ```getAllPermissions()```, calling ```permissions()``` on the user just returning ```MorphToMany``` relationship.

**Reply by erikn69** [reply]
I didn't write `permissions()->`, i did write `permissions->`, read again

**Reply by talelmishali** [reply]
Calling ```permissions``` returning an empty collection.

**Reply by erikn69** [reply]
Calling `permissions` must return all direct permissions for the user, returning to the point, with `auth()->user()` still have you 4 querys?

**Reply by talelmishali** [reply]
I am using permissions via roles. ```getAllPermissions()``` is returning permissions via roles as well. 

Yes, still 4 queries and loading all models twice.

**Reply by erikn69** [reply]
Look at this code, first get the user permissions, then get the roles and permissions through the roles, and combine both, if you know a better way please make a PR, this only happends with `getAllPermissions()`
https://github.com/spatie/laravel-permission/blob/58d5eb6c7b0eafa8bdf0a93d1ca5c214a05344cb/src/Traits/HasPermissions.php#L297-L318

**Comment by talelmishali** [comment]
As mentioned I am sharing ```roles``` and ```permissions``` to the front end with Inertia.js help method ```share```, therefore I instantiate all the ```roles``` and ```permissions``` models to pluck their names.

After that, when accessing an endpoint, let's say ```app.test/clients``` the ```index```method on that controller utilize ```ClientPolicy```, in that policy I use: 
```php
    public function viewAny(User $user)
    {
        if ($user->can('view clients')) {
            return true;
        }
    }
```

Laravel then resolves the ```Gate::class``` contract from the container and will iterate of the user abilities to check if the user does have that permission. In that action Laravel is instantiating all of the permissions models once again, therefore there are twice the models of roles and permissions now instantiated which increases memory use. 

Any good idea how to avoid this?

Thank you,
Tal
